src/serialization/serialization.rs

```diff
@@ -1,3 +1,167 @@
+//! # Serialization/deserialization
+//!
+//! The upstream Java project has established several different types of serialization. We have
+//! currently implemented one (the "V2" format, following the names used by the Java
+//! implementation), and will add others as time goes on. These formats are compact binary
+//! representations of the state of the histogram. They are intended to be used
+//! for archival or transmission to other systems for further analysis. A typical use case would be
+//! to periodically serialize a histogram, save it somewhere, and reset the histogram.
+//!
+//! Histograms are designed to be added, subtracted, and otherwise manipulated, and an efficient
+//! storage format facilitates this. As an example, you might be capturing histograms once a minute
+//! to have a granular view into your performance over time, but you might also want to see longer
+//! trends over an hour or day. Simply deserialize the last 60 minutes worth to recreate their
+//! in-memory `Histogram` form, add them all together into one `Histogram`, and perform whatever
+//! calculations you wish on the resulting histogram. This would allow you to correctly calculate
+//! the 99.99th percentile for the entire hour, for instance, which is not something you can do
+//! if you have only stored percentiles (as opposed to the entire histogram) for each minute.
+//!
+//! # Performance concerns
+//!
+//! Serialization is quite fast; serializing a histogram that represents 1 to `u64::max_value()`
+//! with 3 digits of precision with tens of thousands of recorded counts takes about 40
+//! microseconds on an E5-1650v3 Xeon. Deserialization is about 3x slower, but that will improve as
+//! there are still some optimizations to perform.
+//!
+//! For the V2 format, the space used for a histogram will depend mainly on precision since higher
+//! precision will reduce the extent to which different values are grouped into the same bucket.
+//! Having a large value range (e.g. 1 to `u64::max_value()`) will not directly impact the size if
+//! there are many zero counts as zeros are compressed away.
+//!
+//! # API
+//!
+//! Each serialization format has its own serializer struct, but since each format is reliably
+//! distinguishable from each othere, there is only one `Deserializer` struct that will work for
```

Comment:
Well, this project is a delightful intersection of two things I'm interested in. It pains me how poorly most stacks handle collecting, aggregating, and otherwise processing performance measurements, and while I could go work on revamping that JVM world that I'm quite familiar with, I'm quite enjoying Rust right now, so here I am. I feel like I have an opportunity while Rust is young to get out ahead of people doing it wrong and help make the easy path also be the correct path. :) I enjoy making things bottom-up with the finest quality ingredients, building or improving components if I can't find ones that meet my needs. That's not practical for day-job software generally, but on my own time, I can polish as much as I want...